[Backport FC-0099] feat: add openedx-authz and update libraries enforcement points

MaferMazu opened the pull request:
## Description

Backport of #37564, #37501, #37652, and update requirements.

## Other information

https://github.com/openedx/platform-roadmap/issues/246#issuecomment-2967833993

mariajgrimaldi committed:
feat: filter libraries based on user-role scopes (#37564)

(cherry picked from commit 6c6fc5d55143308f240457ed230a18605ff53ae9)

MaferMazu committed:
feat: add openedx-authz to user_can_create_library and require_permission_for_library_key (#37501)

* feat: add the authz check to the library api function

feat: add the authz publish check in rest_api blocks and containers

feat: add the authz checks in libraries and refactor

feat: add collections checks

feat: update enforcement in serializer file

refactor: refactor the permission check functions

fix: fix value error

fix: calling the queries twice

* test: add structure for test and apply feedback

refactor: refactor the tests and apply feedback

fix: apply feedback

Revert "refactor: refactor the tests and apply feedback"

This reverts commit aa0bd527dd7bc7dec4a7ad7adb41a3c932f4a587.

refactor: use constants and avoid mapping

test: fix the test to have them in order

docs: about we rely on bridgekeeper and the old check for two cases

docs: update openedx/core/djangoapps/content_libraries/api/libraries.py

Co-authored-by: Maria Grimaldi (Majo) <maria.grimaldi@edunext.co>

refactor: use global scope wildcard instead of *

refactor: allow receiving PermissionData objects

refactor: do not inherit from BaseRolesTestCase to favor CL setup methods

If both BaseRolesTestCase and ContentLibrariesRestApiTest define a method
with the same name (e.g., setUp()), Python will use the one found first
in the MRO, which is the one in BaseRolesTestCase because it is
listed first in the class definition leading to unexpected behavior.

refactor: remove unnecessary imports and indent

* chore: bump openedx-authz version

(cherry picked from commit f4f14a69874d5804ef33778486f7d6cab400e206)

openedx-webhooks commented:
Thanks for the pull request, @MaferMazu!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

MaferMazu commented:
@bmtcril, this backport has the consistent Python dependencies workflow failing due to this addition in `edx-lint` https://github.com/openedx/edx-lint/pull/504, but it is not directly related to the changes in this PR. Should we bypass this test when merging?

cc @farhaanbukhsh @mariajgrimaldi

bmtcril commented:
Looks like you should upgrade-package package=<pypi_package_name> to make sure we get a valid dependency tree on the branch.

mariajgrimaldi commented:
Please ensure all of these PRs are included here: https://github.com/search?q=repo%3Aopenedx%2Fedx-platform+FC-0099++is%3Amerged&type=pullrequests. We're currently missing: #37532

MaferMazu commented:
@mariajgrimaldi that PR is already in `release/ulmo` 31b1e6ecc474daabffb778da56c57208788c635b. But thanks, for a moment I thought I forgot.

MaferMazu commented:
Thanks @bmtcril, that fixes the consistency issue.

-----
I'll backport the openedx-authz upgrade, but to maintain consistency, I'll run the package upgrade in this PR, and that will add `Django<6.0` to `requirements/common_constraints.txt` as it happened in this https://github.com/openedx/edx-platform/pull/37633/commits/e8f70b4ce1e8e2f327d83930cebef4718f71f933; that addition is due to this change https://github.com/openedx/edx-lint/pull/504

cc @farhaanbukhsh

MaferMazu commented:
To backport in this PR: https://github.com/openedx/edx-platform/pull/37652

github-actions[bot] committed:
feat: Upgrade Python dependency openedx-authz (#37652)

* feat: Upgrade Python dependency openedx-authz

handle cache invalidation

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

* fix: update the num of queries in tests

---------

Co-authored-by: MaferMazu <35668326+MaferMazu@users.noreply.github.com>
Co-authored-by: Maria Fernanda Magallanes Zubillaga <maria.magallanes@edunext.co>
(cherry picked from commit 122b4e072d6199b51533c29b4fb8e29546d1dc5d)

MaferMazu committed:
chore: update requirements to fix the inconsistency

MaferMazu commented:
This is ready for review @bmtcril @farhaanbukhsh @mariajgrimaldi 🙌

bmtcril reviewed (approved):
👍

mariajgrimaldi reviewed (approved):
I tested this locally alongside tutor@ulmo and tutor-mfe@ulmo and it's working as expected! Thank you so much :)
